zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/callback/ZkAsyncCallbacks.java

```diff
@@ -50,6 +50,10 @@ public void processResult(int rc, String path, Object ctx, byte[] data, Stat sta
         if (_data != null && ctx != null && ctx instanceof ZkAsyncCallMonitorContext) {
           ((ZkAsyncCallMonitorContext) ctx).setBytes(_data.length);
         }
+      } else if(rc != Code.NONODE.intValue()) {
```

Comment:
This logic means, for each retry, we'll record one metric. Is it the expected requirement?
My thought was, only when the whole method fails, we record the failure.

Reply:
This is a good point, thanks. I should perform more tests on this and fix it.

Reply:
This is actually `ZkAsyncCallMonitorContext`, retry won't really go to this path. But I think @rabashizade has other findings regarding retry. He'll update later.

Reply:
Thanks, @zhangmeng916 . Yes, actually due to the catch statement added to asyncSetData by @pkuwm in #1592, the retry and failure paths are different and their monitor contexts are also different. In the case of retry, no failure metric is increased. But a similar check doesn't exist for asyncGetData, and adding that should solve this problem unless there's a reason not to add this logic to asyncGetData.